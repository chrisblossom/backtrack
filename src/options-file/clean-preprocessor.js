/* @flow */

import path from 'path';
import { toArray } from '../utils/object-utils';
import { buildPath, rootPath } from '../config/paths';
import type { Clean, NormalizedClean } from '../types.js';

// eslint-disable-next-line flowtype/require-exact-type
type Args = {
    value?: Clean | $ReadOnlyArray<Clean>,
};

function cleanPreprocessor({ value }: Args = {}): $ReadOnlyArray<
    NormalizedClean,
> {
    const allToArray = toArray(value);

    return allToArray.map((arg: Clean) => {
        const makeDirs = toArray(arg.makeDirs).map((makeDir) => {
            const resolvedMakeDir = path.resolve(buildPath, makeDir);
            return resolvedMakeDir;
        });

        const copy = toArray(arg.copy).map((copyArgs) => {
            const resolvedSrc = path.resolve(rootPath, copyArgs.src);
            const resolvedDest = path.resolve(buildPath, copyArgs.dest);

            const result = {
                src: resolvedSrc,
                dest: resolvedDest,
                hash: copyArgs.hash,
            };

            if (result.hash === undefined) {
                delete result.hash;
            }

            return result;
        });

        return {
            del: toArray(arg.del),
            makeDirs,
            copy,
        };
    });
}

export { cleanPreprocessor };
