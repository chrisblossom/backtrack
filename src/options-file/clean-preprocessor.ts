import path from 'path';
import { toArray } from '../utils/object-utils';
import { buildPath, rootPath } from '../config/paths';
import { Clean, NormalizedClean } from '../types';

interface Args {
	value?: Clean | Clean[];
}

function cleanPreprocessor({ value }: Args = {}): NormalizedClean[] {
	const allToArray = toArray(value);

	return allToArray.map((arg: Clean) => {
		const makeDirs = toArray(arg.makeDirs).map((makeDir) => {
			const resolvedMakeDir = path.resolve(buildPath, makeDir);
			return resolvedMakeDir;
		});

		const copy = toArray(arg.copy).map((copyArgs) => {
			const resolvedSrc = path.resolve(rootPath, copyArgs.src);
			const resolvedDest = path.resolve(buildPath, copyArgs.dest);

			const result = {
				src: resolvedSrc,
				dest: resolvedDest,
				hash: copyArgs.hash,
			};

			if (result.hash === undefined) {
				delete result.hash;
			}

			return result;
		});

		return {
			del: toArray(arg.del),
			makeDirs,
			copy,
		};
	});
}

export { cleanPreprocessor };
