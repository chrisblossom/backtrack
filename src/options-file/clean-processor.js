/* @flow */

import path from 'path';
import { buildPath } from '../config/paths';

import type { NormalizedClean } from '../types.js';

/**
 * Patterns are only allowed inside the buildPath.
 */
const base = {
    del: [],
    makeDirs: [],
};

// eslint-disable-next-line flowtype/require-exact-type
type Args = {
    value: $ReadOnlyArray<NormalizedClean>,
    current?: NormalizedClean,
};

function cleanProcessor({ value, current = base }: Args) {
    return value.reduce(
        (acc, arg) => {
            arg.del.forEach((pattern) => {
                /**
                 * remove duplicates
                 */
                if (acc.del.includes(pattern) === false) {
                    acc.del.push(pattern);
                }
            });

            arg.makeDirs.forEach((dir) => {
                const resolvedDir = path.resolve(buildPath, dir);

                /**
                 * remove duplicates
                 */
                if (acc.makeDirs.includes(resolvedDir) === false) {
                    acc.makeDirs.push(resolvedDir);
                }
            });

            return acc;
        },
        {
            del: [...current.del],
            makeDirs: [...current.makeDirs],
        },
    );
}

export { cleanProcessor };
