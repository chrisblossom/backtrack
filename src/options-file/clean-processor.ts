import { NormalizedClean } from '../types';

/**
 * Patterns are only allowed inside the buildPath.
 */
const base = {
	del: [],
	makeDirs: [],
	copy: [],
};

interface Args {
	value: NormalizedClean[];
	current?: NormalizedClean;
}

function cleanProcessor({ value, current = base }: Args) {
	return value.reduce(
		(acc, arg) => {
			arg.del.forEach((pattern) => {
				/**
				 * remove duplicates
				 */
				if (acc.del.includes(pattern) === false) {
					acc.del.push(pattern);
				}
			});

			arg.makeDirs.forEach((dir) => {
				/**
				 * remove duplicates
				 */
				if (acc.makeDirs.includes(dir) === false) {
					acc.makeDirs.push(dir);
				}
			});

			arg.copy.forEach((copy) => {
				/**
				 * remove duplicates
				 */
				const isDuplicate = !!acc.copy.find((copyArg) => {
					return (
						copyArg.src === copy.src && copyArg.dest === copy.dest
					);
				});

				if (isDuplicate === false) {
					acc.copy.push(copy);
				}
			});

			return acc;
		},
		{
			del: [...current.del],
			makeDirs: [...current.makeDirs],
			copy: [...current.copy],
		},
	);
}

export { cleanProcessor };
