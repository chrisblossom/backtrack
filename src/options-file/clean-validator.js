/* @flow */

import { buildPath } from '../config/paths';
import { fileIsInsideDir } from '../utils/file-is-inside-dir';

import type { NormalizedClean } from '../types.js';

// eslint-disable-next-line flowtype/require-exact-type
type Args = {
    value: $ReadOnlyArray<NormalizedClean>,
};

function cleanValidator({ value }: Args) {
    for (const arg of value) {
        for (const pattern of arg.del) {
            if (fileIsInsideDir(pattern, buildPath) === false) {
                throw new Error(
                    `del pattern '${pattern}' must be inside build directory`,
                );
            }
        }

        for (const dir of arg.makeDirs) {
            if (fileIsInsideDir(dir, buildPath) === false) {
                throw new Error(
                    `makeDirs '${dir}' must be inside build directory`,
                );
            }
        }
    }
}

export { cleanValidator };
