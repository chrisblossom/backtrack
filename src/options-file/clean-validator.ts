import path from 'path';
import { buildPath, rootPath } from '../config/paths';
import { fileIsInsideDirSync } from '../utils/file-is-inside-dir';
import { NormalizedClean } from '../types';

interface Args {
	value: NormalizedClean[];
}

function cleanValidator({ value }: Args): void {
	for (const arg of value) {
		for (const pattern of arg.del) {
			const resolvedPattern = path.resolve(buildPath, pattern);
			if (fileIsInsideDirSync(resolvedPattern, buildPath) === false) {
				throw new Error(
					`del pattern '${pattern}' must be inside build directory`,
				);
			}
		}

		for (const dir of arg.makeDirs) {
			if (fileIsInsideDirSync(dir, buildPath) === false) {
				throw new Error(
					`makeDirs '${dir}' must be inside build directory`,
				);
			}
		}

		for (const copy of arg.copy) {
			if (fileIsInsideDirSync(copy.src, rootPath) === false) {
				throw new Error(
					`copy.src '${copy.src}' must be inside project root directory`,
				);
			}

			if (fileIsInsideDirSync(copy.src, buildPath) === true) {
				throw new Error(
					`copy.src '${copy.src}' cannot be inside build directory`,
				);
			}

			if (fileIsInsideDirSync(copy.dest, buildPath) === false) {
				throw new Error(
					`copy.dest '${copy.dest}' must be inside build directory`,
				);
			}

			const isDuplicateWithHashMismatch = !!arg.copy.find((copyArg) => {
				return (
					copyArg.src === copy.src &&
					copyArg.dest === copy.dest &&
					copyArg.hash !== copy.hash
				);
			});

			if (isDuplicateWithHashMismatch) {
				throw new Error(
					`copy.src '${copy.src}' is duplicated with mismatched hash: true/false`,
				);
			}

			const isDuplicateWithSrcMismatch = !!arg.copy.find((copyArg) => {
				return copyArg.src !== copy.src && copyArg.dest === copy.dest;
			});

			if (isDuplicateWithSrcMismatch) {
				throw new Error(
					`copy.dest '${copy.dest}' is duplicated with mismatched sources`,
				);
			}
		}
	}
}

export { cleanValidator };
