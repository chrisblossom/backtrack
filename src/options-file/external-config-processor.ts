import type { ExternalConfig } from '../types';

interface Args {
	value: Record<string, unknown>;
	current?: ExternalConfig;
}

function externalConfigProcessor({ current, value }: Args): ExternalConfig {
	const result = Object.entries(value).reduce<ExternalConfig>(
		// prettier-ignore
		(acc, [configNamespace, val]: [string, unknown]) => {
			if (acc[configNamespace] === undefined) {
				acc[configNamespace] = [];
			}

			acc[configNamespace].unshift(val);

			return acc;
		},
		{ ...current },
	);

	return result;
}

export { externalConfigProcessor };
