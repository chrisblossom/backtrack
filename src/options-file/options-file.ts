import { loadOptionsFile } from './load-options-file';
import { transformConfig } from './transform-config';
import { rootPath } from '../config/paths';

import { Lifecycles } from '../types';

function optionsFile(dirname: string = rootPath): Lifecycles {
    const baseConfig = loadOptionsFile(dirname);

    const config = transformConfig(baseConfig, dirname);

    return config;
}

export { optionsFile };
