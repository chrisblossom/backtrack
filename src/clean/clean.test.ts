import path from 'path';

const rawClean = (args?: any) => require('./clean').clean(args);
const cleanProcessor = (args: any) =>
	require('../options-file/clean-processor').cleanProcessor(args);
const cleanPreprocessor = (args: any) =>
	require('../options-file/clean-preprocessor').cleanPreprocessor(args);

function clean(files: any) {
	const normalized = cleanPreprocessor({ value: files });
	const parsed = cleanProcessor({ value: normalized });

	return rawClean(parsed);
}

describe('clean', () => {
	const cwd = process.cwd();
	let del = jest.requireMock('del');
	let ensureDir = jest.requireMock('fs-extra').ensureDir;
	let copy = jest.requireMock('fs-extra').copy;

	beforeEach(() => {
		jest.doMock('../utils/log', () => ({
			warn: jest.fn(),
			info: jest.fn(),
			error: jest.fn(),
			success: jest.fn(),
		}));

		jest.doMock('del', () =>
			jest.fn((pattern, options = {}) => {
				const delActual = jest.requireActual('del');

				return delActual(pattern, {
					...options,
					dryRun: true,
				});
			}),
		);

		del = jest.requireMock('del');

		jest.doMock('fs-extra', () => ({
			ensureDir: jest.fn(() => Promise.resolve()),
			copy: jest.fn(() => Promise.resolve()),
		}));

		ensureDir = jest.requireMock('fs-extra').ensureDir;
		copy = jest.requireMock('fs-extra').copy;
	});

	afterEach(() => {
		process.chdir(cwd);
	});

	test('handles undefined', async () => {
		const dir = path.resolve(__dirname, '__sandbox__/1/');
		process.chdir(dir);

		await rawClean();

		expect(del.mock.calls).toMatchSnapshot();
		expect(ensureDir.mock.calls).toEqual([]);
	});

	test('handles one clean', async () => {
		const dir = path.resolve(__dirname, '__sandbox__/1/');
		process.chdir(dir);

		const options = {
			del: ['**/*'],
			makeDirs: [
				'nested/',
				'another_nested/folder',
			],
			copy: [
				{
					src: 'static',
					dest: 'static',
					hash: true,
				},
			],
		};

		await clean(options);

		expect(del.mock.calls).toMatchSnapshot();
		expect(ensureDir.mock.calls).toMatchSnapshot();
		expect(copy.mock.calls).toMatchSnapshot();
	});

	test('handles multiple clean', async () => {
		const dir = path.resolve(__dirname, '__sandbox__/1/');
		process.chdir(dir);

		const options = [
			{
				del: [
					'**/*',
					'!.git',
				],
				makeDirs: [
					'nested/',
					'another_nested/folder',
				],
				copy: [
					{
						src: 'static',
						dest: 'static',
						hash: true,
					},
				],
			},
			{
				del: ['!two.js'],
			},
			{
				makeDirs: ['another_two_nested/folder'],
			},
			{
				copy: [
					{
						src: 'outside-build.js',
						dest: 'static/outside-build.js',
					},
				],
			},
		];

		const normalized = cleanPreprocessor({ value: options });
		const parsed = cleanProcessor({ value: normalized });

		await clean(parsed);

		expect(del.mock.calls).toMatchSnapshot();
		expect(ensureDir.mock.calls).toMatchSnapshot();
		expect(copy.mock.calls).toMatchSnapshot();
	});

	test('handles empty del', async () => {
		const dir = path.resolve(__dirname, '__sandbox__/1/');
		process.chdir(dir);

		const options = [
			{
				makeDirs: [
					'nested/',
					'another_nested/folder',
				],
			},
		];

		await clean(options);

		expect(del.mock.calls).toMatchSnapshot();
		expect(ensureDir.mock.calls).toMatchSnapshot();
	});

	test('handles empty makeDirs', async () => {
		const dir = path.resolve(__dirname, '__sandbox__/1/');
		process.chdir(dir);

		const options = [
			{
				del: ['**/*'],
			},
		];

		await clean(options);

		expect(del.mock.calls).toMatchSnapshot();
		expect(ensureDir.mock.calls).toEqual([]);
	});

	test('merges duplicates in correct order', async () => {
		const dir = path.resolve(__dirname, '__sandbox__/1/');
		process.chdir(dir);

		const options = [
			{
				del: [
					'duplicate/file',
					'unique/file',
					'duplicate/file',
				],
				makeDirs: [
					'duplicate/dir',
					'unique/dir',
					'duplicate/dir',
				],
			},
		];

		await clean(options);

		expect(del.mock.calls).toMatchSnapshot();
		expect(ensureDir.mock.calls).toMatchSnapshot();
	});
});
