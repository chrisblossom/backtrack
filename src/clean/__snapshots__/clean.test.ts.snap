// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clean handles empty del 1`] = `
[
  [
    [],
    {
      "cwd": "<PROJECT_ROOT>/dist",
      "dot": true,
    },
  ],
]
`;

exports[`clean handles empty del 2`] = `
[
  [
    "<PROJECT_ROOT>/dist/nested",
  ],
  [
    "<PROJECT_ROOT>/dist/another_nested/folder",
  ],
]
`;

exports[`clean handles empty makeDirs 1`] = `
[
  [
    [
      "**/*",
    ],
    {
      "cwd": "<PROJECT_ROOT>/dist",
      "dot": true,
    },
  ],
]
`;

exports[`clean handles multiple clean 1`] = `
[
  [
    [
      "**/*",
      "!.git",
      "!two.js",
    ],
    {
      "cwd": "<PROJECT_ROOT>/dist",
      "dot": true,
    },
  ],
]
`;

exports[`clean handles multiple clean 2`] = `
[
  [
    "<PROJECT_ROOT>/dist/nested",
  ],
  [
    "<PROJECT_ROOT>/dist/another_nested/folder",
  ],
  [
    "<PROJECT_ROOT>/dist/another_two_nested/folder",
  ],
]
`;

exports[`clean handles multiple clean 3`] = `
[
  [
    "<PROJECT_ROOT>/outside-build.js",
    "<PROJECT_ROOT>/dist/static/outside-build.js",
    {
      "errorOnExist": true,
      "overwrite": true,
      "preserveTimestamps": true,
    },
  ],
  [
    "<PROJECT_ROOT>/static/static-1.js",
    "<PROJECT_ROOT>/dist/static/static-1.fade7080.js",
    {
      "errorOnExist": true,
      "overwrite": true,
      "preserveTimestamps": true,
    },
  ],
]
`;

exports[`clean handles one clean 1`] = `
[
  [
    [
      "**/*",
    ],
    {
      "cwd": "<PROJECT_ROOT>/dist",
      "dot": true,
    },
  ],
]
`;

exports[`clean handles one clean 2`] = `
[
  [
    "<PROJECT_ROOT>/dist/nested",
  ],
  [
    "<PROJECT_ROOT>/dist/another_nested/folder",
  ],
]
`;

exports[`clean handles one clean 3`] = `
[
  [
    "<PROJECT_ROOT>/static/static-1.js",
    "<PROJECT_ROOT>/dist/static/static-1.fade7080.js",
    {
      "errorOnExist": true,
      "overwrite": true,
      "preserveTimestamps": true,
    },
  ],
]
`;

exports[`clean handles undefined 1`] = `
[
  [
    [],
    {
      "cwd": "<PROJECT_ROOT>/dist",
      "dot": true,
    },
  ],
]
`;

exports[`clean merges duplicates in correct order 1`] = `
[
  [
    [
      "duplicate/file",
      "unique/file",
    ],
    {
      "cwd": "<PROJECT_ROOT>/dist",
      "dot": true,
    },
  ],
]
`;

exports[`clean merges duplicates in correct order 2`] = `
[
  [
    "<PROJECT_ROOT>/dist/duplicate/dir",
  ],
  [
    "<PROJECT_ROOT>/dist/unique/dir",
  ],
]
`;
